internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,606 @@
+	t.Parallel()
+
+	tests := map[string]struct {
+		username  string
+		sessionID string
+		brokerID  string
+
+		secondCall      bool
+		cancelFirstCall bool
+	}{
+		"Successfully authorize":                      {username: "success"},
+		"Successfully authorize after cancel":         {username: "IA_second_call", secondCall: true, cancelFirstCall: true},
+		"Denies authentication when broker times out": {username: "IA_timeout"},
+
+		"Empty data gets JSON formatted": {username: "IA_empty_data"},
+
+		// service errors
+		"Error when sessionID is empty": {sessionID: "empty"},
+		"Error when there is no broker": {sessionID: "no broker", brokerID: "does not exist"},
+
+		// broker errors
+		"Error when authorizing":                            {username: "IA_error"},
+		"Error when broker returns invalid access":          {username: "IA_invalid"},
+		"Error when broker returns invalid data":            {username: "IA_invalid_data"},
+		"Error when calling second time without cancelling": {username: "IA_second_call", secondCall: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			brokerCfgDir := t.TempDir()
+			brokerCfg := testutils.StartBusBrokerMock(t, brokerCfgDir)
+
+			client := startClient(t, brokerCfgDir)
+
+			if tc.brokerID == "" {
+				tc.brokerID = getGeneratedID(t, client, filepath.Base(brokerCfg))
+			}
+
+			if tc.brokerID != "does not exist" {
+				if tc.username == "" {
+					tc.username = "does not matter"
+				}
+				sbResp, err := client.SelectBroker(context.Background(), &authd.SBRequest{
+					BrokerId: tc.brokerID,
+					Username: tc.username,
+				})
+				require.NoError(t, err, "Setup: failed to create session for tests")
+				if tc.sessionID == "" {
+					tc.sessionID = sbResp.GetSessionId()
+				}
+			}
+
+			if tc.sessionID == "empty" {
+				tc.sessionID = ""
+			}
+
+			var firstCall, secondCall string
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			done := make(chan struct{})
+			go func() {
+				defer close(done)
+				iaReq := &authd.IARequest{
+					SessionId:          tc.sessionID,
+					AuthenticationData: "some data",
+				}
+				iaResp, err := client.IsAuthorized(ctx, iaReq)
+				firstCall = fmt.Sprintf("FIRST CALL:\n\taccess: %s\n\tdata: %s\n\terr: %v\n",
+					iaResp.GetAccess(),
+					iaResp.GetData(),
+					err,
+				)
+			}()
+			// Give some time for the first call to block
+			time.Sleep(time.Second)
+			if tc.cancelFirstCall {
+				cancel()
+				// Wait for the cancel to go through
+				time.Sleep(time.Millisecond)
+			}
+
+			if tc.secondCall {
+				iaReq := &authd.IARequest{
+					SessionId:          tc.sessionID,
+					AuthenticationData: "some data",
+				}
+				iaResp, err := client.IsAuthorized(context.Background(), iaReq)
+				secondCall = fmt.Sprintf("SECOND CALL:\n\taccess: %s\n\tdata: %s\n\terr: %v\n",
+					iaResp.GetAccess(),
+					iaResp.GetData(),
+					err,
+				)
+			}
+			<-done
```

Comment:
this <-done should be before if tc.secondCall{} as we should not have two IsAuthorized for the same broker concurrently, no?

Reply:
We shouldn't, yeah. But that's precisely why <-done it's checked later, if it was before we wouldn't be able to test the second call failing because of two concurrent ones.

Reply:
Maybe that’s 2 tests then? (and we reorder the other test to call cancel AFTER the second IsAuthorized call started?)

Reply:
Right, so in `Successfully authorize after cancel`, we want to ensure we cancel, but we don’t necessarily want the first `IsAuthorized` to return before doing the second call, is it right? (the issue here is that it’s racy, and we are not certain about doing this or even IsAuthorized returning before the second is called)

Reply:
Yeah. The previous comment about waiting on `<-done` when we call `cancel()` is a nice change (already done as well, but not pushed). But the second one I think it's how it should be even though we are relying on a timer, it's a very generous one.